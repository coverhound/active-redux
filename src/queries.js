import { createSelector } from 'reselect';
import { apiIndexAsync, apiIndexClear } from './api';

import Registry from './registry';
import Store from './store';

/**
 * @module active-redux/queries
 */

/**
 * Instantiate JSON-API data with its corresponding Model
 * @private
 */
const initModel = (Model) => (data) => data && new Model(data);

/**
 * Map JSON-API data with to corresponding Model
 * @private
 */
const mapModel = (model) => (arr) => arr.map(initModel(model));

const createHash = (type, query) => {
  if (typeof query === 'string') return `${type}.${query}`;

  return Object.entries(query).reduce((hash, [key, value]) => (
    `${hash}.${key}=${value}`
  ), type);
};

const fromIndex = ({ id, type }) => {
  const Model = Registry.get(type);
  return new Model(Store.fromIndex({ id, type }));
};

const createArraySelector = ({ hash, type }) => createSelector(
  (state) => state.api.indices[hash],
  (state) => state.api.resources[type],
  (index = []) => {
    const newIndex = index.map(fromIndex);
    newIndex.isFetching = index.isFetching;
    return newIndex;
  },
);

const createSingleSelector = ({ hash, type }) => createSelector(
  (state) => state.api.indices[hash],
  (state) => state.api.resources[type],
  (index = []) => {
    const newIndex = index.map(fromIndex);
    return newIndex[0];
  },
);

const newPromise = (promise, mapper) => (
  new Promise((resolve, reject) => promise.then(mapper).then(resolve).catch(reject))
);

const wrapPromise = ({ query, array = true, type, promise }) => (dispatch) => {
  const hash = createHash(type, query);
  const selectorCreator = array ? createArraySelector : createSingleSelector;
  const mapper = (array ? mapModel : initModel)(Registry.get(type));
  const wrappedPromise = newPromise(promise, mapper);

  wrappedPromise.selector = selectorCreator({ hash, type });
  wrappedPromise.unsubscribe = () => dispatch(apiIndexClear(hash));

  dispatch(apiIndexAsync({ hash, promise }));

  return wrappedPromise;
};

/**
* Gets all of that resource
* @function
* @example
* Person.all()
* // => Promise<Array<Person>>
* @return {Promise<Array<Model>>} Array of model instances
*/
export const all = (model) => () => (
  wrapPromise({
    query: 'all',
    type: model.type,
    array: true,
    promise: Store.all({ model }),
  })(Store.store.dispatch)
);

/**
* Queries the store for that resource
* @function
* @example
* Person.where({ name: "Joe" })
* // => Promise<Array<Person>>
* @param {Object} query - Query for the store
* @param {Object} [options]
* @param {boolean} [options.remote] - Call to API if no records found?
* @return {Promise<Array<Model>>} Array of model instances
*/
export const where = (model) => (query = {}, options = {}) => (
  wrapPromise({
    query,
    type: model.type,
    array: true,
    promise: Store.where(query, { model, ...options }),
  })(Store.store.dispatch)
);

/**
* Gets one of that resource
* @function
* @example
* Person.find({ id: 5 })
* // => Promise<Person>
* @param {Object} query - Query for the store
* @param {Object} [options]
* @param {boolean} [options.remote] - Call to API if no records found?
* @return {Promise<Model|null>} A model instance
*/
export const find = (model) => (query = {}, options = {}) => (
  wrapPromise({
    query,
    type: model.type,
    array: false,
    promise: Store.find(query, { model, ...options }),
  })(Store.store.dispatch)
);

export const hasMany = (model) => (data) => {
  const id = data.map((d) => d.id);

  return wrapPromise({
    query: { id },
    type: model.type,
    array: true,
    promise: Store.where({ id }, { model })
  })(Store.store.dispatch);
};

export const hasOne = (model) => (data) => {
  const id = data.id;

  return wrapPromise({
    query: { id },
    type: model.type,
    array: false,
    promise: Store.find({ id }, { model })
  })(Store.store.dispatch);
};
