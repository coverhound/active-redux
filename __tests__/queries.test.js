import { bind, define } from 'active-redux';
import * as Queries from 'active-redux/queries';
import mockStore from 'fixtures/store';

bind(mockStore);

describe('Queries', () => {
  const { getState } = mockStore;
  const Person = define('people', class Person {});

  describe('.all()', () => {
    it('returns all of that model', async () => {
      const people = await Queries.all(Person)();
      expect(people).toMatchSnapshot();
    });

    it('s selector returns an index reference', () => {
      const promise = Queries.all(Person)();
      const selector = promise.selector;

      const before = selector(getState());
      expect(before.isFetching).toBe(true);
      expect(before).toMatchSnapshot();

      return promise.then(() => {
        const after = selector(getState());
        expect(after.isFetching).toBe(false);
        expect(after).toMatchSnapshot();
      });
    });
  });

  describe('.where()', () => {
    it('returns some of that model', async () => {
      const people = await Queries.where(Person)({ id: '9' });
      expect(people).toMatchSnapshot();
    });

    it('s selector returns an index reference', () => {
      const promise = Queries.where(Person)({ id: '9' });
      const selector = promise.selector;

      const before = selector(getState());
      expect(before.isFetching).toBe(true);
      expect(before).toMatchSnapshot();

      return promise.then(() => {
        const after = selector(getState());
        expect(after.isFetching).toBe(false);
        expect(after).toMatchSnapshot();
      });
    });
  });

  describe('.find()', () => {
    it('returns the record', async () => {
      const people = await Queries.find(Person)({ id: '9' });
      expect(people).toMatchSnapshot();
    });

    it('s selector returns a single record', () => {
      const promise = Queries.find(Person)({ id: '9' });
      const selector = promise.selector;

      const before = selector(getState());
      expect(before).toBe(undefined);

      return promise.then(() => {
        const after = selector(getState());
        expect(after).toMatchSnapshot();
      });
    });
  });
});
