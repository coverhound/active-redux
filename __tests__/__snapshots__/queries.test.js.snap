// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries .all() returns all of that model 1`] = `
Array [
  Person {
    "data": Object {
      "attributes": Object {
        "address": Object {
          "road": "123 Main St",
        },
        "first-name": "Dan",
        "last-name": "Gebhardt",
        "twitter": "dgeb",
      },
      "id": "9",
      "links": Object {
        "self": "http://example.com/people/9",
      },
      "relationships": Object {
        "comments": Object {
          "data": Array [
            Object {
              "id": "5",
              "type": "comments",
            },
            Object {
              "id": "12",
              "type": "comments",
            },
          ],
        },
      },
      "type": "people",
    },
  },
  Person {
    "data": Object {
      "attributes": Object {
        "address": Object {
          "road": "A different Road",
        },
        "first-name": "Bob",
        "last-name": "Dyland",
        "twitter": "shredmaster",
      },
      "id": "14",
      "links": Object {
        "self": "http://example.com/people/14",
      },
      "type": "people",
    },
  },
]
`;

exports[`Queries .all() s selector returns an index reference 1`] = `Array []`;

exports[`Queries .all() s selector returns an index reference 2`] = `
Array [
  Person {
    "data": Object {
      "attributes": Object {
        "address": Object {
          "road": "123 Main St",
        },
        "first-name": "Dan",
        "last-name": "Gebhardt",
        "twitter": "dgeb",
      },
      "id": "9",
      "links": Object {
        "self": "http://example.com/people/9",
      },
      "relationships": Object {
        "comments": Object {
          "data": Array [
            Object {
              "id": "5",
              "type": "comments",
            },
            Object {
              "id": "12",
              "type": "comments",
            },
          ],
        },
      },
      "type": "people",
    },
  },
  Person {
    "data": Object {
      "attributes": Object {
        "address": Object {
          "road": "A different Road",
        },
        "first-name": "Bob",
        "last-name": "Dyland",
        "twitter": "shredmaster",
      },
      "id": "14",
      "links": Object {
        "self": "http://example.com/people/14",
      },
      "type": "people",
    },
  },
]
`;

exports[`Queries .find() returns the record 1`] = `
Person {
  "data": Object {
    "attributes": Object {
      "address": Object {
        "road": "123 Main St",
      },
      "first-name": "Dan",
      "last-name": "Gebhardt",
      "twitter": "dgeb",
    },
    "id": "9",
    "links": Object {
      "self": "http://example.com/people/9",
    },
    "relationships": Object {
      "comments": Object {
        "data": Array [
          Object {
            "id": "5",
            "type": "comments",
          },
          Object {
            "id": "12",
            "type": "comments",
          },
        ],
      },
    },
    "type": "people",
  },
}
`;

exports[`Queries .find() s selector returns a single record 1`] = `
Person {
  "data": Object {
    "attributes": Object {
      "address": Object {
        "road": "123 Main St",
      },
      "first-name": "Dan",
      "last-name": "Gebhardt",
      "twitter": "dgeb",
    },
    "id": "9",
    "links": Object {
      "self": "http://example.com/people/9",
    },
    "relationships": Object {
      "comments": Object {
        "data": Array [
          Object {
            "id": "5",
            "type": "comments",
          },
          Object {
            "id": "12",
            "type": "comments",
          },
        ],
      },
    },
    "type": "people",
  },
}
`;

exports[`Queries .where() returns some of that model 1`] = `
Array [
  Person {
    "data": Object {
      "attributes": Object {
        "address": Object {
          "road": "123 Main St",
        },
        "first-name": "Dan",
        "last-name": "Gebhardt",
        "twitter": "dgeb",
      },
      "id": "9",
      "links": Object {
        "self": "http://example.com/people/9",
      },
      "relationships": Object {
        "comments": Object {
          "data": Array [
            Object {
              "id": "5",
              "type": "comments",
            },
            Object {
              "id": "12",
              "type": "comments",
            },
          ],
        },
      },
      "type": "people",
    },
  },
]
`;

exports[`Queries .where() s selector returns an index reference 1`] = `Array []`;

exports[`Queries .where() s selector returns an index reference 2`] = `
Array [
  Person {
    "data": Object {
      "attributes": Object {
        "address": Object {
          "road": "123 Main St",
        },
        "first-name": "Dan",
        "last-name": "Gebhardt",
        "twitter": "dgeb",
      },
      "id": "9",
      "links": Object {
        "self": "http://example.com/people/9",
      },
      "relationships": Object {
        "comments": Object {
          "data": Array [
            Object {
              "id": "5",
              "type": "comments",
            },
            Object {
              "id": "12",
              "type": "comments",
            },
          ],
        },
      },
      "type": "people",
    },
  },
]
`;
